### PR TITLE
Stop listing if recursion limit is hit

### LINKED ISSUE #217: gitlab-artifact-cleanup script unable to delete some artifacts
The script worked fine on most projects but not this last one.
Here is the error we get.
powerserver7 / Worklist
Traceback (most recent call last):
  File "/opt/gitlab-artifact-cleanup/gitlab-artifact-cleanup", line 181, in <module>
    main()
------ more logs-----
------ more logs-----
 File "/usr/local/lib/python2.7/dist-packages/python_gitlab-0.15.1-py2.7.egg/gitlab/__init__.py", line 392, in _raw_list
    results.extend(self.list(cls, **args))
  File "/usr/local/lib/python2.7/dist-packages/python_gitlab-0.15.1-py2.7.egg/gitlab/__init__.py", line 468, in list
    return self._raw_list(url, obj_class, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/python_gitlab-0.15.1-py2.7.egg/gitlab/__init__.py", line 392, in _raw_list
    results.extend(self.list(cls, **args))
  File "/usr/local/lib/python2.7/dist-packages/python_gitlab-0.15.1-py2.7.egg/gitlab/__init__.py", line 468, in list
    return self._raw_list(url, obj_class, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/python_gitlab-0.15.1-py2.7.egg/gitlab/__init__.py", line 374, in _raw_list
    r = self._raw_get(path_, **params)
  File "/usr/local/lib/python2.7/dist-packages/python_gitlab-0.15.1-py2.7.egg/gitlab/__init__.py", line 360, in _raw_get
    "Can't connect to GitLab server (%s)" % e)
gitlab.exceptions.GitlabConnectionError: Can't connect to GitLab server (maximum recursion depth exceeded)

And this is our system info.
System information
System:         Ubuntu 14.04
Current User:   git
Using RVM:      no
Ruby Version:   2.3.1p112
Gem Version:    2.6.6
Bundler Version:1.13.6
Rake Version:   10.5.0
Sidekiq Version:4.2.1

GitLab information
Version:        8.14.4
Revision:       3ea6c8d
Directory:      /opt/gitlab/embedded/service/gitlab-rails
DB Adapter:     postgresql
URL:            	
HTTP Clone URL: 	
SSH Clone URL:  	
Using LDAP:     yes
Using Omniauth: no

GitLab Shell
Version:        4.0.3
Repository storage paths:
- default:      /var/opt/gitlab/git-data/repositories
Hooks:          /opt/gitlab/embedded/service/gitlab-shell/hooks/
Git:            /opt/gitlab/embedded/bin/git

Please help

### PR DESCRIPTION
**Pull request by johanbrandhorst**
Fixes #217

### CODE REVIEW

**Comment by gpocentek** [comment]
Thank you for working on this, hitting the server limits is an annoying problem.

With your patch using ``all=True`` might return incomplete results, without any information given to the user. I'm not really fond of this behavior. Maybe we should give the user a choice: return a partial list, raise an exception (default, to keep the current behavior), or maybe wait a few seconds before continuing (see #166). What do you think?

**Comment by johanbrandhorst** [comment]
Good point on incomplete results, I agree this should change to keep current behaviour. This patch does not concern itself with rate limiting though, just the python built-in recursion limit. I can add an option to return as many as possible (and capture the recursion limit exception), while maintaining the same behaviour for `all=True`.

**Comment by gpocentek** [comment]
Oh right, I was focused on the API rate limit!

Adding the option would be great, thanks :)

File: gitlab/__init__.py [hunk 1]
```diff
@@ -112,8 +112,8 @@ def __init__(self, url, private_token=None, email=None, password=None,
         # build the "submanagers"
         for parent_cls in six.itervalues(globals()):
             if (not inspect.isclass(parent_cls)
-               or not issubclass(parent_cls, GitlabObject)
-               or parent_cls == CurrentUser):
+                or not issubclass(parent_cls, GitlabObject)
+                    or parent_cls == CurrentUser):
```
Comment (johanbrandhorst) [inline gitlab/__init__.py]: My `autopep8` linter did this, should I revert it?

Comment (gpocentek) [inline gitlab/__init__.py]: No need to revert, that's fine.

**Comment by gpocentek** [comment]
This looks good to me, thank you for adding unit tests!

Could you have a look at the travis job failures? Thanks :)

**Comment by johanbrandhorst** [comment]
Should fix the python2 errors

**Comment by gpocentek** [comment]
I fixed the pep8 issues, thank you for this patch!

**Pull request merged**